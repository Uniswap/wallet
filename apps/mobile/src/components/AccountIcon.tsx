import React from 'react'
import { StyleSheet } from 'react-native'
import Svg, { Defs, RadialGradient as RadialGradientSVG, Rect, Stop } from 'react-native-svg'
import { Unicon } from 'src/components/unicons/Unicon'
import { useUniconColors } from 'src/components/unicons/utils'
import { Flex, Icons } from 'ui/src'
import { spacing } from 'ui/src/theme'
import { RemoteImage } from 'wallet/src/features/images/RemoteImage'

export interface AccountIconProps {
  size: number
  showViewOnlyBadge?: boolean
  viewOnlyBadgeScalingFactor?: number
  address: string
  avatarUri?: string | null
  showBackground?: boolean // Display images with solid background.
}

const INSET_PADDING = spacing.spacing16

export function AccountIcon({
  size,
  showViewOnlyBadge,
  address,
  avatarUri,
  showBackground,
  viewOnlyBadgeScalingFactor = 0.45,
}: AccountIconProps): JSX.Element {
  // If background, add padding and center Unicons. Leave ENS avatars as is.
  const shouldShowUniconInsetPadding = !avatarUri && showBackground

  // If Unicon and background, reduce size to account for added padding.
  const adjustedIconSize = shouldShowUniconInsetPadding ? size - INSET_PADDING * 2 : size

  // Color for gradient background.
  const { gradientStart: uniconColor } = useUniconColors(address)

  const iconPadding = size * 0.15
  const iconEyeContainerSize = size * viewOnlyBadgeScalingFactor
  const iconEyeSize = iconEyeContainerSize - iconPadding

  const defaultImage = (
    <>
      <Unicon address={address} size={adjustedIconSize} />
      {showBackground ? <UniconGradient color={uniconColor} size={size} /> : null}
    </>
  )

  return (
    <Flex
      backgroundColor={showBackground ? '$surface1' : '$transparent'}
      borderColor={showBackground ? '$surface1' : '$transparent'}
      borderRadius="$roundedFull"
      borderWidth={showBackground ? 2 : 0}
      position="relative"
      style={{
        padding: shouldShowUniconInsetPadding ? INSET_PADDING : spacing.none,
      }}>
      {avatarUri ? (
        <RemoteImage
          borderRadius={adjustedIconSize}
          fallback={defaultImage}
          height={adjustedIconSize}
          uri={avatarUri}
          width={adjustedIconSize}
        />
      ) : (
        defaultImage
      )}
      {showViewOnlyBadge && (
        <Flex
          alignItems="center"
          backgroundColor="$surface2"
          borderRadius="$roundedFull"
          bottom={-2}
          height={iconEyeContainerSize}
          justifyContent="center"
          position="absolute"
          right={-2}
          shadowColor="$sporeBlack"
          shadowOffset={{ width: 0, height: 0 }}
          shadowOpacity={0.2}
          shadowRadius={10}
          width={iconEyeContainerSize}>
          <Icons.Eye color="$neutral1" size={iconEyeSize} />
        </Flex>
      )}
    </Flex>
  )
}

// Circle shaped gradient that follows Unicon colors.
const UniconGradient = ({ color, size }: { color: string; size: number }): JSX.Element => {
  return (
    <Svg height={size} style={UniconGradientStyles.svg} width={size}>
      <Defs>
        <RadialGradientSVG cy="-0.1" id="background" rx="0.8" ry="1.1">
          <Stop offset="0" stopColor={color} stopOpacity="0.6" />
          <Stop offset="1" stopColor={color} stopOpacity="0" />
        </RadialGradientSVG>
      </Defs>
      <Rect
        fill="url(#background)"
        height="100%"
        opacity={0.6}
        rx={size}
        width="100%"
        x="0"
        y="0"
      />
    </Svg>
  )
}

const UniconGradientStyles = StyleSheet.create({
  svg: {
    position: 'absolute',
  },
})
