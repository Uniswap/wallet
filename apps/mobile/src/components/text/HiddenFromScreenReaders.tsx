import React, { PropsWithChildren } from 'react'
import { View } from 'react-native'

export function HiddenFromScreenReaders<T>({ children }: PropsWithChildren<T>): JSX.Element {
  return (
    <View accessibilityElementsHidden={true} importantForAccessibility="no-hide-descendants">
      {children}
    </View>
  )
}
