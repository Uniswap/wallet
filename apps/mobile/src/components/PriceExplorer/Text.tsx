import React from 'react'
import { SharedValue, useAnimatedStyle, useDerivedValue } from 'react-native-reanimated'
import { useLineChartDatetime } from 'react-native-wagmi-charts'
import { useAppTheme } from 'src/app/hooks'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { AnimatedText } from 'src/components/text/AnimatedText'
import { AnimatedDecimalNumber } from './AnimatedDecimalNumber'
import { useLineChartPrice, useLineChartRelativeChange } from './usePrice'

export function PriceText({ loading }: { loading: boolean }): JSX.Element {
  const price = useLineChartPrice()

  if (loading) {
    return <Text loading loadingPlaceholderText="$10,000" variant="headlineLarge" />
  }

  return <AnimatedDecimalNumber number={price} testID="price-text" variant="headlineLarge" />
}

export function RelativeChangeText({
  loading,
  spotRelativeChange,
}: {
  loading: boolean
  spotRelativeChange?: SharedValue<number>
}): JSX.Element {
  const theme = useAppTheme()

  const relativeChange = useLineChartRelativeChange({ spotRelativeChange })
  const icon = useDerivedValue(() => (relativeChange.value.value > 0 ? '↗' : '↘'))
  const styles = useAnimatedStyle(() => ({
    color:
      relativeChange.value.value > 0 ? theme.colors.accentSuccess : theme.colors.accentCritical,
  }))

  if (loading) {
    return <Text loading loadingPlaceholderText="00.00%" variant="bodyLarge" />
  }

  return (
    <Flex row gap="spacing2">
      <AnimatedText style={styles} testID="relative-change-icon" text={icon} variant="bodyLarge" />
      <AnimatedText
        style={styles}
        testID="relative-change-text"
        text={relativeChange.formatted}
        variant="bodyLarge"
      />
    </Flex>
  )
}

export function DatetimeText({ loading }: { loading: boolean }): JSX.Element | null {
  // `datetime` when scrubbing the chart
  const datetime = useLineChartDatetime()

  if (loading) return null

  return <AnimatedText color="textSecondary" text={datetime.formatted} variant="bodyLarge" />
}
