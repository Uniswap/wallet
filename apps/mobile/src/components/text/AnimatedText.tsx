import {
  color,
  ColorProps,
  createRestyleComponent,
  createVariant,
  typography,
  TypographyProps,
  useResponsiveProp,
  VariantProps,
} from '@shopify/restyle'
import React from 'react'
import {
  TextInput,
  TextInputProps,
  TextProps as RNTextProps,
  useWindowDimensions,
} from 'react-native'
import Animated, { useAnimatedProps } from 'react-native-reanimated'
import { DEFAULT_FONT_SCALE } from 'src/components/Text'
import { textVariants } from 'ui/src/theme/restyle/font'
import { Theme } from 'ui/src/theme/restyle/theme'

// base animated text component using a TextInput
// forked from https://github.com/wcandillon/react-native-redash/blob/master/src/ReText.tsx
Animated.addWhitelistedNativeProps({ text: true })

interface TextProps extends Omit<TextInputProps, 'value' | 'style'> {
  text: Animated.SharedValue<string>
  style?: Animated.AnimateProps<RNTextProps>['style']
}

const AnimatedTextInput = Animated.createAnimatedComponent(TextInput)
export const BaseAnimatedText = (props: TextProps): JSX.Element => {
  const { style, text, ...rest } = props
  const animatedProps = useAnimatedProps(() => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return {
      text: text.value,
      // Here we use any because the text prop is not available in the type
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any
  })
  return (
    <AnimatedTextInput
      editable={false}
      style={[style || undefined]}
      underlineColorAndroid="transparent"
      value={text.value}
      {...rest}
      {...{ animatedProps }}
    />
  )
}
// end of forked from https://github.com/wcandillon/react-native-redash/blob/master/src/ReText.tsx

// exposes restyle to base animated text
const StyledBaseAnimatedText = createRestyleComponent<
  VariantProps<Theme, 'textVariants'> &
    TypographyProps<Theme> &
    ColorProps<Theme> &
    React.ComponentProps<typeof BaseAnimatedText>,
  Theme
>([createVariant({ themeKey: 'textVariants' }), typography, color], BaseAnimatedText)

// wrapped around restyled animated text with convenience props
export const AnimatedText = (
  props: React.ComponentProps<typeof StyledBaseAnimatedText>
): JSX.Element => {
  const { fontScale } = useWindowDimensions()
  const enableFontScaling = fontScale > DEFAULT_FONT_SCALE

  const variant = useResponsiveProp(props.variant ?? 'bodySmall') as keyof typeof textVariants
  const multiplier = textVariants[variant].maxFontSizeMultiplier

  return (
    <StyledBaseAnimatedText
      {...props}
      allowFontScaling={enableFontScaling}
      maxFontSizeMultiplier={multiplier}
    />
  )
}
