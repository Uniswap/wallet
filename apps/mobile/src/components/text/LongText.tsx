/* eslint-disable react-native/no-inline-styles */
import React, { ComponentProps, useCallback, useReducer, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { LayoutChangeEvent, NativeSyntheticEvent, TextLayoutEventData } from 'react-native'
import Markdown from 'react-native-markdown-display'
import { useAppTheme } from 'src/app/hooks'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { openUri } from 'src/utils/linking'

type LongTextProps = {
  initialDisplayedLines?: number
  text: string
  gap?: keyof Theme['spacing']
  color?: string
  linkColor?: string
  readMoreOrLessColor?: string
  renderAsMarkdown?: boolean
  variant?: keyof Theme['textVariants']
} & Omit<
  ComponentProps<typeof Text>,
  'children' | 'numberOfLines' | 'onTextLayout' | 'color' | 'variant'
>

export function LongText(props: LongTextProps): JSX.Element {
  const theme = useAppTheme()
  const { t } = useTranslation()
  const {
    initialDisplayedLines = 3,
    text,
    gap = 'spacing8',
    color = theme.colors.textPrimary,
    linkColor = theme.colors.accentAction,
    readMoreOrLessColor = theme.colors.accentAction,
    renderAsMarkdown = false,
    variant = 'bodySmall',
    ...rest
  } = props

  const [expanded, toggleExpanded] = useReducer(
    (isExpanded) => !isExpanded,
    renderAsMarkdown ? true : false
  )
  const [textLengthExceedsLimit, setTextLengthExceedsLimit] = useState(false)
  const [initialContentHeight, setInitialContentHeight] = useState<number | undefined>(undefined)

  const textLineHeight = theme.textVariants[variant].lineHeight
  const maxVisibleHeight = textLineHeight * initialDisplayedLines

  const onLayout = useCallback(
    (event: LayoutChangeEvent) => {
      if (!renderAsMarkdown || initialContentHeight !== undefined) return
      const textContentHeight = event.nativeEvent.layout.height
      setTextLengthExceedsLimit(textContentHeight > maxVisibleHeight)
      toggleExpanded()
      setInitialContentHeight(textContentHeight)
    },
    [initialContentHeight, maxVisibleHeight, renderAsMarkdown]
  )

  const onTextLayout = useCallback(
    (e: NativeSyntheticEvent<TextLayoutEventData>) => {
      if (!renderAsMarkdown) {
        setTextLengthExceedsLimit(e.nativeEvent.lines.length >= initialDisplayedLines)
      }
    },
    [renderAsMarkdown, initialDisplayedLines]
  )

  return (
    <Flex gap={gap}>
      <Text
        numberOfLines={expanded ? undefined : initialDisplayedLines}
        style={
          renderAsMarkdown
            ? {
                color,
                height: !textLengthExceedsLimit || expanded ? 'auto' : maxVisibleHeight,
                overflow: 'hidden',
              }
            : { color }
        }
        variant={variant}
        onLayout={onLayout}
        onTextLayout={onTextLayout}
        {...rest}>
        {renderAsMarkdown ? (
          <Markdown
            style={{
              body: { color },
              link: { color: linkColor },
              paragraph: {
                marginBottom: 0,
                marginTop: 0,
                fontSize: theme.textVariants.bodySmall.fontSize,
                lineHeight: theme.textVariants.bodySmall.lineHeight,
              },
            }}
            onLinkPress={(url): false => {
              // add our own custom link handler since it has security checks that only open http/https links
              openUri(url)
              return false
            }}>
            {text}
          </Markdown>
        ) : (
          text
        )}
      </Text>

      {/* Text is removed vs hidden using opacity to ensure spacing after the element is consistent in all cases.
      This will cause mild thrash as data loads into a page but will ensure consistent spacing */}
      {textLengthExceedsLimit ? (
        <Text
          my="none"
          py="none"
          style={{ color: readMoreOrLessColor }}
          testID="read-more-button"
          variant="buttonLabelSmall"
          onPress={toggleExpanded}>
          {expanded ? t('Read less') : t('Read more')}
        </Text>
      ) : null}
    </Flex>
  )
}
