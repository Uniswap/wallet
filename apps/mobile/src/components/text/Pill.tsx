import React, { ReactNode } from 'react'
import { Flex, FlexProps } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'

type PillProps = {
  customBackgroundColor?: string
  customBorderColor?: string
  foregroundColor?: string
  icon?: ReactNode
  label?: ReactNode
  textVariant?: keyof Theme['textVariants']
} & FlexProps

export function Pill({
  borderRadius = 'roundedFull',
  children,
  customBackgroundColor,
  customBorderColor,
  flexDirection = 'row',
  foregroundColor,
  icon,
  label,
  px = 'spacing4',
  py = 'spacing8',
  textVariant = 'bodySmall',
  ...rest
}: PillProps): JSX.Element {
  return (
    <Flex
      alignItems="center"
      borderColor="none"
      borderRadius={borderRadius}
      borderWidth={1}
      flexDirection={flexDirection}
      gap="spacing8"
      justifyContent="center"
      px={px}
      py={py}
      style={{
        ...(customBackgroundColor ? { backgroundColor: customBackgroundColor } : {}),
        ...(customBorderColor ? { borderColor: customBorderColor } : {}),
      }}
      {...rest}>
      {icon ?? null}
      {label ? (
        <Text style={{ color: foregroundColor }} variant={textVariant}>
          {label}
        </Text>
      ) : null}
      {children}
    </Flex>
  )
}
