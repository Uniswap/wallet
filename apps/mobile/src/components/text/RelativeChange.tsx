import React from 'react'
import { useAppTheme } from 'src/app/hooks'
import { Caret } from 'src/components/icons/Caret'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { formatNumber, NumberType } from 'src/utils/format'

interface RelativeChangeProps {
  change?: number
  absoluteChange?: number
  variant?: keyof Theme['textVariants']
  semanticColor?: boolean // If true, entire % change text will render green or red
  positiveChangeColor?: keyof Theme['colors']
  negativeChangeColor?: keyof Theme['colors']
  arrowSize?: number
  loading?: boolean
  alignRight?: boolean
}

export function RelativeChange(props: RelativeChangeProps): JSX.Element {
  const theme = useAppTheme()
  const {
    absoluteChange,
    change,
    variant = 'subheadSmall',
    semanticColor,
    positiveChangeColor = 'accentSuccess',
    negativeChangeColor = 'accentCritical',
    arrowSize = theme.iconSizes.icon16,
    loading = false,
    alignRight = false,
  } = props

  const isPositiveChange = change !== undefined ? change >= 0 : undefined
  const arrowColor = isPositiveChange
    ? theme.colors[positiveChangeColor]
    : theme.colors[negativeChangeColor]

  const formattedChange = change !== undefined ? `${Math.abs(change).toFixed(2)}%` : '-'
  const formattedAbsChange = absoluteChange
    ? `${formatNumber(Math.abs(absoluteChange), NumberType.PortfolioBalance)}`
    : ''

  return (
    <Flex
      row
      alignItems="center"
      gap="spacing2"
      justifyContent={alignRight ? 'flex-end' : 'flex-start'}>
      {change !== undefined && (
        <Caret color={arrowColor} direction={isPositiveChange ? 'n' : 's'} size={arrowSize} />
      )}
      <Text
        color={
          semanticColor ? (isPositiveChange ? 'accentSuccess' : 'accentCritical') : 'textSecondary'
        }
        loading={loading}
        loadingPlaceholderText="$0.00 (0.00)%"
        variant={variant}>
        {absoluteChange ? `${formattedAbsChange} (${formattedChange})` : formattedChange}
      </Text>
    </Flex>
  )
}
