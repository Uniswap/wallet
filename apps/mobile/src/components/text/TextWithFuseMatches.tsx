import Fuse from 'fuse.js'
import React from 'react'
import { TextProps } from 'react-native'
import { Text } from 'src/components/Text'
import { Theme } from 'src/styles/theme'
import { Flex } from '../layout'

interface TextWithFuseMatchesProps {
  text: string
  matches?: readonly Fuse.FuseResultMatch[]
  variant?: keyof Theme['textVariants']
  numberOfLines?: TextProps['numberOfLines']
}

export function TextWithFuseMatches({
  matches,
  text,
  variant = 'bodyLarge',
  numberOfLines = 1,
}: TextWithFuseMatchesProps & TextProps): JSX.Element {
  if (!matches || matches.length === 0) {
    return (
      <Text color="textPrimary" numberOfLines={numberOfLines} variant={variant}>
        {text}
      </Text>
    )
  }

  const charIsMatch = new Set()
  for (const match of matches) {
    for (const index of match.indices) {
      for (let i = index[0]; i < index[1] + 1; i++) {
        charIsMatch.add(i)
      }
    }
  }

  // PERF: batch pieces?
  const pieces = []
  for (let i = 0; i < text.length; i++) {
    if (charIsMatch.has(i)) {
      pieces.push([text[i], true])
    } else {
      pieces.push([text[i], false])
    }
  }

  const elements = (
    <>
      {pieces.map((p, i) => {
        if (p[1])
          return (
            <Text key={`${i}${p[0]}`} color="textPrimary" variant={variant}>
              {p[0]}
            </Text>
          )
        else
          return (
            <Text key={`${i}${p[0]}`} color="textTertiary" variant={variant}>
              {p[0]}
            </Text>
          )
      })}
    </>
  )

  return <Flex row>{elements}</Flex>
}
