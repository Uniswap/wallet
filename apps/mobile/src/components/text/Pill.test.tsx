import React from 'react'
import { Text } from 'src/components/Text'
import { Pill } from 'src/components/text/Pill'
import { render } from 'src/test/test-utils'

it('renders a Pill without image', () => {
  const tree = render(
    <Pill backgroundColor="background1" foregroundColor="accentActive" label="My Pill Label" />
  )
  expect(tree).toMatchSnapshot()
})

it('renders a Pill with border', () => {
  const tree = render(
    <Pill borderColor="accentSuccess" icon={<Text>Icon</Text>} label="My Second Pill Label" />
  )
  expect(tree).toMatchSnapshot()
})
