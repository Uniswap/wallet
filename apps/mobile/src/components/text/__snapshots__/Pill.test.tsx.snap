// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Pill with border 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderColor": "#40B66B",
        "borderRadius": 999999,
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 4,
        "paddingVertical": 8,
      },
      Object {},
    ]
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      Array [
        Object {
          "display": "flex",
        },
      ]
    }
  >
    Icon
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "width": 8,
        },
      ]
    }
  />
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      Array [
        Object {
          "color": "#0E1524",
          "display": "flex",
          "fontFamily": "Inter-Regular",
          "fontSize": 15,
          "lineHeight": 20,
        },
        Object {
          "color": undefined,
        },
      ]
    }
  >
    My Second Pill Label
  </Text>
</View>
`;

exports[`renders a Pill without image 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderRadius": 999999,
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 4,
        "paddingVertical": 8,
      },
      Object {},
    ]
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      Array [
        Object {
          "color": "#0E1524",
          "display": "flex",
          "fontFamily": "Inter-Regular",
          "fontSize": 15,
          "lineHeight": 20,
        },
        Object {
          "color": "accentActive",
        },
      ]
    }
  >
    My Pill Label
  </Text>
</View>
`;
